### ISSUE TITLE
Gameplan for evolving new UI

### ISSUE THREAD

**Issue by littleguy77** [mupen64plus-ae/mupen64plus-ae#420]
@BonzaiThePenguin 

Your work in the New-Gallery-UI branch looks fantastic.  I really like the use of standard design idioms and re-use of standardized components.  Rather than just merging your branch when it's finished, I'd like to propose an incremental process for integrating this into the master branch, that should minimize regressions and allow incremental testing.

I would like to begin manually cherry-picking your work onto the master branch, starting with the support libraries (appcompat, gridlayout, recyclerview).  I'll break those commits up a bit for atomicity and add some extra help docs (particularly the README.md), to allow other devs to start using them right away (mainly thinking about @xperia64 who said he wanted to update the cheats editor UI).  I'll tweak the app manifest and related stuff as well to be sure everything is consistent.

After that, I will start cherry-picking your work, starting with foundational stuff like resources and preferences.  Then I'll start picking the actual implementation, providing code review as I go.  If I notice any bugs or design conflicts we can discuss as I go.  Since you are brand new to the project, and the existing code isn't always documented well, I think this review process will help to minimize regressions.

If this approach sounds good to you, then I suggest you keep hammering away on the New-Gallery-UI branch as the pathfinder prototype, with the idea that features will be pulled over to master as they stabilize.  Don't worry about making "perfect" commits on that branch, since I can clean them up when I cherry-pick to master.  (It should go without saying that you will retain author attribution whenever I cherry-pick.)

**Comment by littleguy77** [comment]
@BonzaiThePenguin I made a test branch with the appcompat stuff and updated readme file and looks to be working fine for me.  Once @paulscode updates his auto-build script, I'll start pushing to the master branch and give you feedback on any commits if necessary, and testers will be able to start testing.  Everything I've seen so far looks great :+1:

**Comment by BonzaiThePenguin** [comment]
Will we be removing the AppCompat stuff from the libs folder when that happens? That would really simplify the diffs, which right now are flooded with those files.

**Comment by littleguy77** [comment]
Yes, I was going to break out the app compat stuff into separate commits before pushing to master.  You can see a dry-run here:
https://github.com/littleguy77/mupen64plus-ae/commits/new-gallery-ui

Though I will probably rebase/tweak a few more times before it finally goes to master.

**Comment by littleguy77** [comment]
I found a few more calls elsewhere that require API 11.  Very easy for me to spot since one of my test devices is an Xperia Play (API 9).

Once I get the appcompat stuff into master, I'll start cherry-picking your commits and we can go through them step by step (and fold fixups into them as we go).

**Comment by BonzaiThePenguin** [comment]
The Android SDK came with an Android API 9 emulator, so once I finish up these last few things I'll run it through that.

**Comment by BonzaiThePenguin** [comment]
All fixed. The search field isn't showing up for me though, not sure what's up with that.

**Comment by littleguy77** [comment]
I'm thinking we might be chasing our tails a bit on the refresh process.  I feel like the use case was ironed out very well and we got positive feedback testers.  It just needed some UI tweaks.  Now I feel like the UI tail is wagging the dog on it.

From my understanding, the only issue with the old approach was related to screen real estate.  I was thinking a logical evolution would be to just make it an activity (or better yet, a fragment).  The activity would have a button to spawn our standard file/folder dialog, three checkboxes, and an OK/Cancel.  Let me get the foundational stuff refactored and onto master, and then I'll layout (or prototype) my ideas.  Until then, let's focus on the other stuff like context menus and flow between activities.

I hate to slow you down or throw water on the fire, but I also don't want to have tunnel vision.  We can avoid tunnel vision by getting more devs/testers in the loop, which will happen as soon as stuff starts getting into the autobuilds.  Which is what I'm really trying to focus on this weekend.

**Comment by BonzaiThePenguin** [comment]
The new system actually works the same as the old one – you click the button, choose a file or folder to add to the library, and it adds it. The only difference was that "search zips" was global rather than per-folder, and now that's the same.

But consider these situations:

---

You're trying to clear up space on your SD card, so you go through your games folders and delete games you haven't played in a while. Later on you notice the ROMs are still listed in the Mupen library and there's no easy way to get rid of them. Even if there was a Remove option (with all of the fun permissions issues that come with that territory) that'd still mean the user would have to go through and remove the missing files manually.

But that's of course why there's the "Clear gallery" option. You click refresh ROMs, select a folder, check "Clear gallery", click OK, and half your library disappears because some of your ROMs were stored in a different folder! Logically that system only makes sense is if they only ever use one folder, but we previously established that it needs to support adding files from multiple folders. Ultimately that means they have to manually re-add every ROMs folder one-by-one every time they want to get clear out files that no longer exist.

With the new system you just click the "Update library" icon.

---

Then of course there's the classic situation of getting a new ROM and wanting to add it to your library. So you click "update ROMs", navigate to the folder, and either select the file or the folder containing it. With the new system you still just click the "Update library" icon, and it will find it automatically without making you navigate through the filesystem every time.

In file managers this sync feature is called "watch folders", and has always been highly requested in the managers I've made due to how much easier it makes things for the user. Mupen64Plus creates these watch folders as you add folders of files to your library, so the next time you make any changes in the file system you only need to tap the update button and the app will take care of everything. But the old functionality is still there too – you can add the file manually using the (+) button.

---

I probably should have brought it up beforehand, but the file management system seemed like much more of a placeholder to me than the gallery layout!

**Comment by BonzaiThePenguin** [comment]
Regarding the appcompat stuff, we could just include a shell script that configures the support libraries properly. No need to make them do it manually. It really just needs to fix the manifest and project properties for gridlayout and recyclerview.

(Not sure why that didn't occur to me sooner.)

**Comment by littleguy77** [comment]
Thanks, your explanations are very helpful.  I think the last time I tested the new refresh process some things were broken (like the clearing) and may be fixed now, hence the confusion.

I'm not opposed to keeping the appcompat stuff inside the repository, if it makes it dead simple for new contributors and auto-build scripts.  But obviously many projects have dealt with this already, so it might be good to see how they did it.  My main focus for now was to just separate the commits so that your actual changes were easier to see.

I'm going to crank through this today, even if the auto-build stuff isn't working yet.

**Comment by littleguy77** [comment]
Also, I forget about SD card speed since I don't have or use them on most of my test devices.  That's a very good point about the searching.

Ok enough chatter from me I need to get to work catching up! :D

**Comment by BonzaiThePenguin** [comment]
I did a lot of research on the appcompat stuff earlier today and determined that they give directions for Eclipse and Android Studio only specifically *because* these projects aren't configured properly, and their command-line tools aren't up to snuff. It sounds crazy, but:

The android tool doesn't understand absolute paths to libraries, so you need to calculate a relative path from your project to the SDK's extras/android/support/v7 folder. *From the command-line*, where there's no built-in tool to do so.
http://stackoverflow.com/questions/18209041/how-to-define-v7-appcompat-dependency-correctly

The android tool also has a bug with "update lib-project", in that it fails to write the fact that it's a library to the project.properties file. The only way to set this flag is to create a brand-new library project and import all of the files into it.
https://code.google.com/p/android/issues/detail?id=38270

It also fails to remove the activity embedded within the application for the recyclerview's manifest file.

So what's the solution? Use a computer program that has a plugin that fixes these problems for you! Once I realized that's why Google recommends we use Eclipse or Android Studio, it suddenly occurred to me that writing a shell script to do the same thing would work just as well.

So if you want I can make a shell script that documents this stuff and handles the configuration for the projects, then we can just ask them to download the Support Library as part of their SDK download.

**Comment by BonzaiThePenguin** [comment]
Another option is to switch from ant to gradle. That came up a lot as a recommendation, but if you want to keep ant then we'd need to include the properly-configured appcompat libraries in some fashion.

**Comment by littleguy77** [comment]
Take a peek at master now -- I just pushed the appcompat stuff.  You'll notice I only pushed the jar file for recyclerview, because that is a "library with no resources".  So that takes care of excluding the recyclerview's manifest.

If you are using Eclipse, just re-import the mupen64plus-ae root folder and it should detect the new locations of appcompat and gridviewmanager.

I think when the dust settles we can explore new build processes, but for now the current system is working very easily from a fresh git clone, and I'd like to minimize the number of parallel issues for the moment.

**Comment by littleguy77** [comment]
**Important**

Ok, I just rebased your branch onto master.  You should do a force update of your local working copy.  Assuming you are already in the New-Gallery-UI branch, this would be
```
# Stash any uncommitted changes
git stash
# Get all the latest commits, removing branches that no longer exist on GitHub
git fetch --prune --all
# Hard-reset to the latest tip of the new gallery branch
get reset --hard origin/New-Gallery-UI
# Re-apply any uncommitted changes
git stash pop
```

**Comment by littleguy77** [comment]
I have been stepping through this one commit at a time.  The implementation looks solid.  I'm squashing fixups and interposing my own fixes on this temporary branch:
https://github.com/littleguy77/mupen64plus-ae/commits/new-ui-rev1

As you can see I'm clumping the commits by feature.

**Comment by BonzaiThePenguin** [comment]
The `git stash pop` step failed and got me stuck in some weird middle state. I hate git. May as well just delete the folder and clone it into the GitHub app again.

**Comment by littleguy77** [comment]
Are you using the GitHub app?  Because it sucks big time.  I'd recommend SourceTree or GitExtensions.  The former is a little simpler (my work colleagues say more noob friendly), the latter is a little more powerful (my personal favorite).

**Comment by BonzaiThePenguin** [comment]
I only use the GitHub app to clone and commit changes. It turned out the app also seems to support guiding me through the git stash problems, but I didn't notice until after I had already chosen to remove the repo. Saw out of the corner of my eye something that said "Having git stash problems? Don't worry! We'll..." aaaaand I deleted it. Whoops.

**Comment by littleguy77** [comment]
Sorry about that :(

**Comment by BonzaiThePenguin** [comment]
> Take a peek at master now -- I just pushed the appcompat stuff. You'll notice I only pushed the jar file for recyclerview, because that is a "library with no resources". So that takes care of excluding the recyclerview's manifest.

...oh snap, of course! In that case we really only need to calculate a relative path to appcompat and gridlayout, and link directly to the jar for recyclerview. No need to reconfigure anything. We should definitely consider looking into that once the dust settles on this merge.

**Comment by littleguy77** [comment]
> We should definitely consider looking into that once the dust settles on this merge.

Sure.

BTW just to confirm, you are able to build and test with the latest on master, correct?  It should be as simple as following the README if you have a fresh clone.

**Comment by BonzaiThePenguin** [comment]
I don't use Eclipse, but did you want me to set it up to make sure your steps work?

**Comment by littleguy77** [comment]
No, that's fine, I tested already a bunch of times myself.

I won't try to convert you to an IDE, but eclipse does make it easy to spot all sorts of minSDK issues (that's how I caught a few already).

**Comment by BonzaiThePenguin** [comment]
Eclipse helped me find a ton of unused packages and a few unused variables the last time I tried it, which was quite useful. Only reason I stick with the command-line is because I'm low on hard drive space!

**Comment by littleguy77** [comment]
Gotta hit the sack.  Looking great so far.  The one major issue so far is this:
https://github.com/littleguy77/mupen64plus-ae/commit/9b10e6a45cc

One minor nit, could we make the gallery-item context dots a material-style button (like the main one in the upper-right corner) so that when you tap it you get instant visual feedback?

**Comment by BonzaiThePenguin** [comment]
I just successfully compiled by using a relative link to the appcompat and gridlayout libraries. I had to call `android update lib-project -p <path to library>` on each of them first, though.

> One minor nit, could we make the gallery-item context dots a material-style button (like the main one in the upper-right corner) so that when you tap it you get instant visual feedback?

Yes of course, that was bothering me too. All you need to do is set the background drawable on it, and there are a ton of built-in drawables to choose from. The one I chose for the other widgets gives things an orange background.

**Comment by BonzaiThePenguin** [comment]
> The one major issue so far is this:
littleguy77@9b10e6a

The app doesn't use that code anymore. It was replaced with a drawer that slides out from the left.

**Comment by BonzaiThePenguin** [comment]
You're going to have a *bad* time if you test every commit against older APIs – I didn't worry about API compatibility until it was feature-complete. The latest version was tested and works on API 9. Also quite a few features that were added ended up being removed before all was said and done, and some features that were ripped out were slowly re-implemented.

**Comment by littleguy77** [comment]
Fair enough, then it is ok not to implement a backwards compatible solution.  But every commit should still at least build and run without crashing older devices.  If something is incompatible with older APIs then it must be surrounded with a guard and the limitation noted in a comment.

**Comment by BonzaiThePenguin** [comment]
>> One minor nit, could we make the gallery-item context dots a material-style button (like the main one in the upper-right corner) so that when you tap it you get instant visual feedback?

> Yes of course, that was bothering me too. All you need to do is set the background drawable on it, and there are a ton of built-in drawables to choose from. The one I chose for the other widgets gives things an orange background.

Correction: I was using other Material-themed apps (YouTube, Play Store, etc.) and I noticed that the only visual feedback they provide when you're pushing the ... button is that the underlying list item is not selected. This matches the behavior in this app, but the part I was doing wrong is that it shouldn't show the sidebar on a long press – only long pressing the list item should do that.

**Comment by BonzaiThePenguin** [comment]
Maybe we should just choose a more muted color for selecting list items – I only used orange because that just so happened to be the first one I tried, and I was only trying to get keyboard navigation *functional*, not pretty. :P

Speaking of colors, you can change any of the colors in the UI if you want, like the Floating Action Button. Those reds are literally identical to the ones that appear when you visit [http://www.w3schools.com/tags/ref_colorpicker.asp](the w3schools HTML color picker). My only goal was to find a color that made it not-so-obvious that the FAB doesn't have a shadow (since that would have taken more code and I was in no mood to implement it).

**Comment by BonzaiThePenguin** [comment]
The latest commit is the experimental stuff. In particular I'm worried that since I futzed around with the view hierarchy for the emulator, I may have destroyed the rendering performance. Really need to test that out before I work on the actual logic for the sidebar.

**Comment by littleguy77** [comment]
Cool.  Last night I started playing around with the new gallery and must say I really love the side drawers.  Much better use of space IMO than the action bar in many situations.  It did occur to me that using swipe gestures to open drawers in the game activity might cause issues with the touchscreen controls.  E.g. user is mashing virtual buttons and accidentally slides open a drawer.  Probably easiest just to build it and see.

On another note I'm finding it a bit difficult to touch the gallery item context menu button.  Long-press also seems flaky for some reason -- it doesn't always seem to detect.  Then I found myself accidentally touching a grid item and launching the game when I didn't want to (which then put it in the recent games list, where I didn't want it).  It then occurred to me, why don't we just invert the press/long-press relationship?  Just idea to toss out there:
 - Eliminate context menu dots on gallery item
 - On long-click, jump straight to GameActivity (resume)
 - On normal-click, open a game-specific drawer on the right-hand side, with a menu structure similar to current PlayMenuActivity (and thus eliminating PlayMenuActivity)
    - Resume
    - Restart
    - Multiplayer mapping (if multiplayer game and multiple controllers are enabled)
    - Emulation profile
    - Touchscreen profile
    - Controller x profile
    - ...
    - Cheats
    - Wiki
    - Defaults

There are a lot of things I like about that setup:
 - The game specific settings feature is obvious for users coming from version 2.x (not everyone thinks to long-click)
 - It's not a big deal if a user never discovers the long-click feature (it's just icing on the cake)
 - It eliminates the PlayMenuActivity
 - I like the mental distinction of the drawers (left is global, right is game-specific)

The right-hand drawer would also be a good place to put a future "Saves" entry, which could jump to a new (TBD) SaveDataActivity where users can browse a list of save states for a game, where each is annotated with a screenshot.  Tapping the save state entry launches the game and immediately loads the selected state.

**Comment by BonzaiThePenguin** [comment]
Are you using the latest version, or one of the middle commits? I made the hit target much larger near the end of development.

**Comment by littleguy77** [comment]
Sorry, should have been clear.  I am talking about the current WIP pull request.  I did see your fix to enlarge, and folded it into the appropriate commit in that PR.  I think part of the issue may be the change from dip to pixels, since my gallery items are very small now.

In any case, I think there's still good food for thought on the proposition I just made about inverting the long-click behavior.

**Comment by BonzaiThePenguin** [comment]
> It did occur to me that using swipe gestures to open drawers in the game activity might cause issues with the touchscreen controls. E.g. user is mashing virtual buttons and accidentally slides open a drawer. Probably easiest just to build it and see.

The other two/three edges already trigger UI elements with an edge swipe, with the back/home/task bar and the status bar. I wanted to put the game sidebar on the right side, but the right side is taken up by that home button bar when in landscape view.

> I think part of the issue may be the change from dip to pixels, since my gallery items are very small now.

Not noticing that on my end.

> In any case, I think there's still good food for thought on the proposition I just made about inverting the long-click behavior.

It would definitely be nice to get rid of the PlayMenuActivity entirely, and your idea for getting rid of the dots seems good. I secretly didn't like single taps going directly to gameplay anyway. Long-presses could be a useful metaphor for "PLAY THIS NOW". Plus yeah it would clean up the interface a lot.

**Comment by BonzaiThePenguin** [comment]
How do you profile rendering performance and RAM usage? I can't tell if the emulation is running any slower with the new sidebar, but I am noticing that my CPU monitor (the tiny white line at the top of my screenshots on paulscode.com) keeps quitting in the middle of playing a game, and when I quit Mupen my home launcher has to relaunch from scratch. It didn't used to do that.

But then again I haven't actually played a game on here since before work on New-Gallery-UI started, so maybe something else is causing it or my memory is bad.

**Comment by littleguy77** [comment]
Yeah I'm on a 7 inch tablet (2012 nexus 7) and am seeing (3 rows x 6 columns) in landscape and (5 rows by 4 columns) in portrait.  Each grid is less than 1 square inch in real life -- which looks very nice IMO.  Just a little small if there's a context menu.

> How do you profile rendering performance and RAM usage? 

FPS can be displayed by ticking the box in the global settings.  That's my bottom-line standard.  I must admit that I rarely look at RAM usage, and when I do I just use random apps from Google Play to get the gist (nothing rigorous).

**Comment by BonzaiThePenguin** [comment]
Thanks, I'll give it a try.

Also, I just got rid of the dots view and swapped the long-press behavior, and suddenly the thumbnails are the correct size again! Looks like that was the problem...

**Comment by littleguy77** [comment]
@BonzaiThePenguin 
Master now includes the new grid mechanics, gallery activity drawer, and a few new options.
https://github.com/mupen64plus-ae/mupen64plus-ae/commits/master

Looking through your New-Gallery-UI branch, I see a lot of distinct features/enhancements.  For manageability, could you split these commits into separate logical branches?  Each branch should start at https://github.com/mupen64plus-ae/mupen64plus-ae/commit/a54a1a03a70098ab2d82d6b9a0bc25fc497d704a

My priorities for getting stuff into master
 - Tweaks to navigation drawer in GalleryActivity
      - Indentation, menu structure
      - Language should be in the first level of the hierarchy so people can find it
 - Replace PlayMenuActivity with nav drawer
      - Nav drawer opens when you touch/click gallery item (let's pretend the dots menu never existed)
      - Contains all the same functionality of PlayMenuActivity
      - Cover art at the top is a great touch :)
      - *Do not add the rip info yet* - just stick with replacing PlayMenuActivity, no more on this branch
 - Enhance ROM scan process
      - New UI to address real estate issues of scan_roms_dialog
      - Per-folder zip feature (I still haven't been able to see this feature because I get crashes)
      - The default folder should *not* be `!Downloads` (part of the crash issue, I think), just `!`
 - GameActivity navigation drawer
      - Sounds there's a lot of tweaking an polishing here, so definitely on its own branch
 - Rom info enhancements
      - These are nice but also create a cascade of work to translators and upstream, hence we should isolate these on their own branch
 - Other new features (?)
      - Material Design-ification of all activities
      - Crash reporting
      - Cheats

**Comment by BonzaiThePenguin** [comment]
> Replace PlayMenuActivity with nav drawer

I'm still working on that part. This is actually a massive change due to how much functionality is packed into PlayMenuActivity. Profiles, resetting defaults, Wiki link, and the entire cheats editor.

> Rom info enhancements
These are nice but also create a cascade of work to translators and upstream, hence we should isolate these on their own branch

I removed that a few days ago, in favor of simpler one-word headings that match the text on the sort options. So like instead of "Released in USA: Game was released in the USA and Japan", it just says "Region: USA/Japan" or whatever. And then you can sort by "Region".

> The default folder should not be !Downloads (part of the crash issue, I think), just !

I changed this a long time ago, and also fixed the crash associated with it. Come on! :P

**Comment by littleguy77** [comment]
The latest still crashes for me, for whatever reason. :(

**Comment by littleguy77** [comment]
It's best to put them in their own branches even if (precisely because) they're not finished.  Right now everything is in various states of completion but they're all dependent upon each other.  I want to get them decoupled for sanity's sake.

**Comment by BonzaiThePenguin** [comment]
> The latest still crashes for me, for whatever reason. :(

Doesn't the debugger tell you the function and line that caused the crash?

**Comment by littleguy77** [comment]
Actually, I can't even build the head of New-Gallery-UI.  I'm getting a build error I've never seen before:
[2015-03-14 11:01:34 - mupen64plus-ae] Dx 1 error; aborting
[2015-03-14 11:01:34 - mupen64plus-ae] Conversion to Dalvik format failed with error 1

If you start a clean branch that just includes the work you've done for the replacement to the PlayMenuActivity, then I can probably finish, test, and merge it to master.  In the process I can migrate to a config-file (rather than SharedPreferences) approach for all game-specific settings.

**Comment by BonzaiThePenguin** [comment]
Here's what I could find:

> This is often due to inadvertently including a core library file in your application's project, when using an IDE (such as Eclipse).

http://stackoverflow.com/questions/11537975/android-dx-error1-failed-to-convert-to-dalvik-format

**Comment by BonzaiThePenguin** [comment]
I'm still compiling from the command-line, by running android update lib-project and fixing the project.properties file.

**Comment by littleguy77** [comment]
Thanks.  It was actually due to some junk left over by Dropbox when I unsynced some unrelated stuff.

I was able to get your latest working after I deleted the GalleryCache directory and reset all global settings to their defaults.  The last time I tried that must have been before you fixed the default folder location.  I also had to deal with some minSDK issues in the code before it would build.  Anyhow, it works great!

I would still like to keep all these changes isolated from each other for the sake of regression testing and sanity.

**Comment by BonzaiThePenguin** [comment]
> I was able to get your latest working after I deleted the GalleryCache directory and reset all global settings to their defaults.

Hrm, I was hoping to fix the crash instead. :P Otherwise it's just going to pop up again with someone else. It's probably caused by the folder path not existing, so I can test it by adding a watch folder then deleting the folder via a file manager.

> I would still like to keep all these changes isolated from each other for the sake of regression testing and sanity.

Not sure when I'll get started on that, maybe Tuesday or Wednesday.

**Comment by BonzaiThePenguin** [comment]
(the crash I fixed was due to the path not containing a forward slash character)

**Comment by littleguy77** [comment]
> Not sure when I'll get started on that, maybe Tuesday or Wednesday.

The sooner the better.  I would **much** rather get what we have under control on separate branches before making more changes (even if they are bugfixes).  Really, that is the highest priority at this point.

**Comment by littleguy77** [comment]
Ok, looks like I got us into a sticky situation and I take full responsibility for the mess I have created and my poor communication.  I spent a good chunk of time yesterday trying to rebase New-Gallery-UI onto the head of master but the best I could do was this:
https://github.com/mupen64plus-ae/mupen64plus-ae/commits/new-ui-rebase1

Beyond that it just got too messy to deconflict my changes from yours.

I'm not sure where to go from here, but I'm strongly considering rewinding master back to the New-Gallery-UI branch point.  Maybe from there we can decide on a feature at a time to merge and we can shift gears from "exploration" to "integration".
